From 8a63398b0bcbaebab163ec44f47212931c40d0d3 Mon Sep 17 00:00:00 2001
From: Arie Brown <abrown@habana.ai>
Date: Tue, 30 Nov 2021 18:13:50 +0200
Subject: [PATCH] [SW-63523] fuse: reverse fields in serialize

Add is_reverse parameter to version serialize function as preparation
for gaudi2 printing fuse version.

Change-Id: I8a197947eeddd0a13c05b27296fad0687d0cbbfc
Signed-off-by: Arie Brown <abrown@habana.ai>
---

diff --git a/app/preboot/preboot.c b/app/preboot/preboot.c
index 42ffabc..cbf3a24 100644
--- a/app/preboot/preboot.c
+++ b/app/preboot/preboot.c
@@ -197,7 +197,8 @@
 };
 
 void preboot_fuse_version_serialize(char *ver, size_t size,
-				    struct preboot_fuse_ver_data *ver_data)
+				    struct preboot_fuse_ver_data *ver_data,
+				    bool is_reverse)
 {
 	struct preboot_str_size ver_str[] = {
 		{ ver_data->table_ver, ver_data->table_ver_sz },
@@ -208,13 +209,14 @@
 		{ ver_data->xaddr, ver_data->xaddr_sz },
 		{ ver_data->yaddr, ver_data->yaddr_sz },
 	};
-	int i, j;
+	int i, j, str_index;
 
 	for (i = 0; i < ARRAY_SIZE(ver_str); i++) {
 		for (j = 0; j < ver_str[i].size; j++) {
 			if (!size)
 				break;
-			*ver++ = ver_str[i].str[j];
+			str_index = is_reverse ? ver_str[i].size - 1 - j : j;
+			*ver++ = ver_str[i].str[str_index];
 			size--;
 		}
 		if (!size)
diff --git a/arch/gaudi/arm/asic.c b/arch/gaudi/arm/asic.c
index 9988f53..c3729a3 100644
--- a/arch/gaudi/arm/asic.c
+++ b/arch/gaudi/arm/asic.c
@@ -175,7 +175,7 @@
 
 	memcpy(hl_asic_sn, fuse.dev_id, sizeof(hl_asic_sn));
 
-	preboot_fuse_version_serialize(version, size, &ver_data);
+	preboot_fuse_version_serialize(version, size, &ver_data, false);
 
 	return 0;
 }
diff --git a/include/preboot.h b/include/preboot.h
index 9ebdc00..8543dbc 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -181,7 +181,8 @@
 };
 
 void preboot_fuse_version_serialize(char *ver, size_t size,
-				    struct preboot_fuse_ver_data *fuse_data);
+				    struct preboot_fuse_ver_data *fuse_data,
+				    bool is_reverse);
 bool preboot_sram_is_valid_address_range(u64 addr, int len);
 bool preboot_is_valid_address_range(u64 addr, int len);
 bool preboot_is_overlap_address_range(u64 addr, int len);
